ambry-account/src/main/java/com/github/ambry/account/RouterStore.java

```diff
@@ -277,15 +284,33 @@ public ZNRecord update(ZNRecord znRecord) {
         accountServiceMetrics.accountUpdateToAmbryTimeInMs.update(System.currentTimeMillis() - startTimeMs);
       } catch (Exception e) {
         accountServiceMetrics.accountUpdatesToAmbryServerErrorCount.inc();
-        errorMessage =
-            "Updating accounts failed because unexpected error occurred when uploading AccountMetadata to ambry";
-        logger.error(errorMessage, e);
-        throw new IllegalStateException(errorMessage, e);
+        logAndThrowIllegalStateException(
+            "Updating accounts failed because unexpected error occurred when uploading AccountMetadata to ambry", e);
       }
 
       // Start step 5:
-      accountBlobIDs.add(new BlobIDAndVersion(this.newBlobID, newVersion).toJson());
-      recordToUpdate.setListField(ACCOUNT_METADATA_BLOB_IDS_LIST_KEY, accountBlobIDs);
+      if (blobIDAndVersions.size() + 1 > totalNumberOfVersionToKeep) {
+        Iterator<BlobIDAndVersion> iter = blobIDAndVersions.iterator();
+        while (blobIDAndVersions.size() + 1 > totalNumberOfVersionToKeep) {
+          BlobIDAndVersion blobIDAndVersion = iter.next();
+          iter.remove();
+          logger.info("Removing blob " + blobIDAndVersion.getBlobID() + " at version " + blobIDAndVersion.getVersion());
+          try {
+            router.get().deleteBlob(blobIDAndVersion.getBlobID(), SERVICE_ID).get();
+          } catch (Exception e) {
+            logger.error("Failed to delete blob={} from older version because of {}", blobIDAndVersion.getBlobID(), e);
+            accountServiceMetrics.accountDeletesToAmbryServerErrorCount.inc();
+          }
+        }
+        // Clear json string list since it's not sorted.
+        blobIDAndVersionsJson.clear();
```

Comment:
Could you just do `blobIDAndVersionsJson = blobIDAndVersions.stream().map(BlobIDAndVersion::toJson).collect(Collectors.toList())`?